### ISSUE TITLE
SC2154 not reported if referenced in arbitrary condition

### ISSUE THREAD

**Issue by kkmuffme** [koalaman/shellcheck#3117]
#### For bugs
- Rule Id: SC2154
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#! /bin/bash

if [[ -n "$x" ]]
then
        echo "$bar"
fi

echo "$bar"
```

Also for easy-to-make mistakes when refactoring:

```
if [[ -z "${bar+x}" ]]
then
    # bar is not defined
    echo "$bar"
fi
```

#### Here's what shellcheck currently says:

>No issues detected!

#### Here's what I wanted or expected to see:

>Line 8 SC2154 (warning): bar is referenced but not assigned.
